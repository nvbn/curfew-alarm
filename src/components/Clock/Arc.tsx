import React from "react";
import { Path } from "react-native-svg";

const polarToCartesian = (
  cx: number,
  cy: number,
  r: number,
  d: number,
): [number, number] => {
  const dRadians = ((d - 90) * Math.PI) / 180;

  return [cx + r * Math.cos(dRadians), cy + r * Math.sin(dRadians)];
};

// Based on https://stackoverflow.com/a/18473154
const makeArc = (
  cx: number,
  cy: number,
  r: number,
  start: number,
  end: number,
): string => {
  const adjustedEnd = end < start ? end + 360 : end;

  const [startX, startY] = polarToCartesian(cx, cy, r, adjustedEnd); // swap?
  const [endX, endY] = polarToCartesian(cx, cy, r, start);

  const largeArcFlag = adjustedEnd - start <= 180 ? 0 : 1;

  return `
    M ${startX} ${startY}
    A ${r} ${r} 0 ${largeArcFlag} 0 ${endX} ${endY}
  `;
};

type Props = {
  startMinute: number;
  endMinute: number;
  color: string;
  width: number;
  cx: number;
  cy: number;
  r: number;
};

/**
 * SVG arc from `startMinute` to `endMinute`.
 */
const Arc = ({
  startMinute,
  endMinute,
  color,
  width,
  cx,
  cy,
  r,
}: Props): JSX.Element => (
  <Path
    d={makeArc(cx, cy, r, startMinute / 4, endMinute / 4)}
    stroke={color}
    fill="none"
    strokeWidth={width}
  />
);

export default Arc;
