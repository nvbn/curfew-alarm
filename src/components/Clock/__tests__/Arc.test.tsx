import { render } from "@testing-library/react-native";
import React from "react";
import Arc from "../Arc";

describe("<Arc />", () => {
  test("10pt red arc from 12:00 to 14:45", () => {
    const rendered = render(
      <Arc
        startMinute={12 * 60}
        endMinute={14 * 60 + 45}
        color="red"
        width={10}
        cx={0}
        cy={0}
        r={10}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  test("15pt cyan arc from 21:00 to 04:30", () => {
    const rendered = render(
      <Arc
        startMinute={21 * 60}
        endMinute={4 * 60 + 30}
        color="cyan"
        width={15}
        cx={10}
        cy={10}
        r={15}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
