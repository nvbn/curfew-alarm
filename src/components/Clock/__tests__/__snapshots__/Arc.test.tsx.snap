// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Arc /> 10pt red arc from 12:00 to 14:45 1`] = `
<RNSVGPath
  d="
    M -6.593458151000688 7.518398074789774
    A 10 10 0 0 0 6.123233995736766e-16 10
  "
  fill={null}
  propList={
    Array [
      "fill",
      "stroke",
      "strokeWidth",
    ]
  }
  stroke={4294901760}
  strokeWidth={10}
/>
`;

exports[`<Arc /> 15pt cyan arc from 21:00 to 04:30 1`] = `
<RNSVGPath
  d="
    M 23.858192987669305 4.259748514523657
    A 15 15 0 0 0 -0.606601717798215 -0.6066017177982115
  "
  fill={null}
  propList={
    Array [
      "fill",
      "stroke",
      "strokeWidth",
    ]
  }
  stroke={4278255615}
  strokeWidth={15}
/>
`;
